gprofiler/proc_events.py

```diff
@@ -0,0 +1,129 @@
+from select import select
+import errno
+import struct
+import socket
+import threading
+import os
+
+from gprofiler.log import get_logger_adapter
+
+logger = get_logger_adapter(__name__)
+
+
+# linux/netlink.h:
+_NETLINK_CONNECTOR = 11
+
+# linux/netlink.h:
+_NLMSG_DONE = 0x3  # End of a dump
+
+# struct nlmsghdr {
+#         __u32           nlmsg_len;      /* Length of message including header */
+#         __u16           nlmsg_type;     /* Message content */
+#         __u16           nlmsg_flags;    /* Additional flags */
+#         __u32           nlmsg_seq;      /* Sequence number */
+#         __u32           nlmsg_pid;      /* Sending process port ID */
+# };
+_nlmsghdr = struct.Struct("=I2H2I")
+
+
+# linux/connector.h:
+_CN_IDX_PROC = 0x1
+
+# struct cn_msg {
+#         struct cb_id id;
+#
+#         __u32 seq;
+#         __u32 ack;
+#
+#         __u16 len;              /* Length of the following data */
+#         __u16 flags;
+#         __u8 data[0];
+# };
+_cn_msg = struct.Struct("=4I2H")
+
+
+# linux/cn_proc.h:
+# struct proc_event {
+#         enum what {
+#                 ...
+#         } what;
+#         __u32 cpu;
+#         __u64 __attribute__((aligned(8))) timestamp_ns;
+#                 /* Number of nano seconds since system boot */
+#         union { /* must be last field of proc_event struct */
+#                 ...
+#         } event_data;
+# };
+_base_proc_event = struct.Struct("=2IQ")
+
+# From enum what
+_PROC_EVENT_EXIT = 0x80000000
+
+# struct exit_proc_event {
+#         __kernel_pid_t process_pid;
+#         __kernel_pid_t process_tgid;
+#         __u32 exit_code, exit_signal;
+# } exit;
+_exit_proc_event = struct.Struct("=4I")
+
+
+def _proc_events_listener():
```

Comment:
I'd wrap it in a class with proper lifetime management (`__enter__`, `__exit__` etc)

Reply:
Thread control, ...

Reply:
What're the advantages you see over the current approach?
The way I see it: you'll never want more than a single thread to listen to connector messages, and you would never want it to die. Hence there is no meaning for creating or holding an object to represent this thread, and there is no reason to use OOP when you don't need an object.
In this case exporting just a bunch of `register_.*_callback` functions is a much clearer and simpler API.

Reply:
Summary of my verbal discussion with @Jongy: We might want to restart the thread during execution, hence `__enter__` and `__exit__` functionality will be useful.